unstructured_inference/models/unstructuredmodel.py

```diff
@@ -55,6 +64,99 @@ def __call__(self, x: Image) -> List[LayoutElement]:
+            for table in tables:
+                nested_current = np.array([e.is_almost_subregion_of(table) for e in not_tables])
+                if nested_check is None:
+                    nested_check = nested_current
+                    continue
+                nested_check = nested_check | nested_current
+
+            final_elements = []
+            for nested, elem in zip(nested_check, not_tables):  # type:ignore
+                if not nested:
+                    final_elements.append(elem)
+
+            final_elements.extend(tables)
+            final_elements.sort(key=lambda e: e.y1)
+            return final_elements
+
+        def enhance_regions(
+            elements: List[LayoutElement],
+            min_text_size: int,
+            iom_to_merge: float = 0.3,
+        ) -> List[LayoutElement]:
+            """This function traverses all the elements and either deletes nested elements,
+            or merges or splits them depending on the iou score for both regions"""
+            intersections_mtx = intersections(*elements)
+
+            for i, row in enumerate(intersections_mtx):
+                first = elements[i]
+                if first:
+                    # We get only the elements which intersected
+                    indices_to_check = np.where(row)[0]
+                for j in indices_to_check:
+                    if i != j and elements[j] is not None and elements[i] is not None:
+                        second = elements[j]
+                        intersection = first.intersection(
+                            second,
+                        )  # we know it does, but need the region
+                        first_inside_second = first.is_in(second)
+                        second_inside_first = second.is_in(first)
+
+                        if first_inside_second and not second_inside_first:
+                            elements[i] = None  # type:ignore
+                        elif second_inside_first and not first_inside_second:
+                            # delete second element
+                            elements[j] = None  # type:ignore
+                        elif intersection:
+                            iom = first.intersection_over_minimum(second)
+                            if iom < iom_to_merge:  # small
+                                separate(first, second)
+                                # The rectangle could become too small, which is a
+                                # good size to delete?
+                            else:  # big
+                                # merge
+                                if first.area > second.area:
+                                    grow_region_to_match_region(first, second)
+                                    elements[j] = None  # type:ignore
+                                else:
+                                    grow_region_to_match_region(second, first)
+                                    elements[i] = None  # type:ignore
```

Comment:
so because this inner loop just goes down the list of all overlapped boxes we could get boxes (lets call it `j+k`) that completely contains this one (`i`) later but another box (current `j`) would have intersection over minimum so we grow `j` while the most appropriate action might be simply dedupe `i` (and later `j` likely would be folded into `j+k` isntead.
It feels like we should first compute the intersection area matrix between pairs of boxes and go from there to merge. That way we know the full picture before changing any elements.

Reply:
Updated to cover that case